x-pack/platform/plugins/shared/observability_solution/observability_ai_assistant/server/utils/recall/rewrite_user_prompt.ts

```diff
@@ -0,0 +1,303 @@
+    properties: {
+      field: { const: field },
+      value: { enum: values },
+    },
+    required: ['field', 'value'],
+  }));
+
+  const REWRITE_USER_PROMPT_FUNCTION_NAME = `rewrite_user_prompt-${indexPattern}`;
+
+  const rewriteUserPromptFunction = {
+    strict: true,
+    name: REWRITE_USER_PROMPT_FUNCTION_NAME,
+    description: `Rewrites the user prompt into a concise search query and infers term filters and date range filter to be used for filtering the dataset in Elasticsearch.
+
+          ## Example
+          User prompt: "List all the bug reports that were opened for the front page this week?" 
+          Screen description: "User is looking at the front page of the website"
+
+          """
+          {
+            "rewrittenUserPrompt": "front page bugs",
+            "keywordFilters": [
+              { "field": "type", "value": "issue" }
+            ],
+            "timestampFilter": {
+              { "field": "@timestamp", "gte": "now-7d/d", "lte": "now/d" }
+            }
+          }
+          """
+        `,
+    parameters: {
+      type: 'object',
+      properties: {
+        rewrittenUserPrompt: {
+          type: 'string',
+          description:
+            "A single sentence that captures the core of the user's intent. The rewritten user prompt will be used as a semantic search query.",
+        },
+        ...(keywordFieldSchema.length > 0
+          ? {
+              keywordFilters: {
+                type: 'array',
+                description:
+                  'A list of Elasticsearch filters for exactly matching keyword fields. The filters are used to narrow down the search results based on user intent.',
+                items: {
+                  type: 'object',
+                  description:
+                    'A filter object for a named keyword field, filtered by a specific value.',
+                  oneOf: keywordFieldSchema,
+                },
+              } as const,
+            }
+          : {}),
+        ...(timestampFields.length > 0
+          ? {
+              timestampFilter: {
+                type: 'object',
+                description: 'A range filter to specify a date range.',
+                properties: {
+                  field: {
+                    type: 'string',
+                    description: 'The name of the date field',
+                    enum: timestampFields,
+                  },
+                  gte: {
+                    type: 'string',
+                    description:
+                      "The greater-than-or-equal value (e.g., a date string like 'now-7d/d').",
+                  },
+                  lte: {
+                    type: 'string',
+                    description: "The less-than-or-equal value (e.g., a date string like 'now/d').",
+                  },
+                },
+                required: ['field', 'gte', 'lte'],
+                additionalProperties: false,
+              } as const,
```

Comment:
@arturoliduena Right now we have a very strict schema. I like that. The downside is that the LLM does not generate   ES query filters but instead just plain objects, which we have convert to ES query filters.